scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java

```diff
@@ -682,67 +721,29 @@ private OpRequest inferOpRequest(Type type, String name, Map<TypeVariable<?>, Ty
 		return new DefaultOpRequest(name, type, mappedOutputs[0], mappedInputs);
 	}
 
-	private synchronized void initOpDirectory() {
-		if (opDirectory != null) return;
-		Map<String, List<OpInfo>> tmp = new HashMap<>();
-		// add all OpInfos that are directly discoverable
-		discoverers.stream().flatMap(d -> d.discover(OpInfo.class).stream()).forEach(info -> addToOpIndex.accept(tmp, info));
-		List<OpInfoGenerator> generators = infoGenerators();
-		discoverers.stream().flatMap(d -> d.discover(Op.class).stream()).forEach(o -> registerOpsFrom(tmp, o, generators));
-		discoverers.stream().flatMap(d -> d.discover(OpCollection.class).stream()).forEach(o -> registerOpsFrom(tmp, o, generators));
-		Set<OpInfo> infos = tmp.values().stream().flatMap(Collection::stream).map(info -> opsFromObject(info, generators)).flatMap(Collection::stream).collect(
-				Collectors.toSet());
-		infos.forEach(info -> addToOpIndex.accept(tmp, info));
-		opDirectory = tmp;
-	}
-
-	/**
-	 * Generates a {@link List} of {@link OpInfo}s from {@code o} using a List of {@link OpInfoGenerator}s.
-	 * @param o the {@link Object} to parse {@link OpInfo}s from.
-	 * @param generators the {@link List} of {@link OpInfoGenerator}s
-	 * @return a {@link List} of {@link OpInfo}s
-	 */
-	private List<OpInfo> opsFromObject(Object o, List<OpInfoGenerator> generators) {
-		return generators.stream() //
-				.filter(g -> g.canGenerateFrom(o)) //
-				.flatMap(g -> g.generateInfosFrom(o).stream()) //
-				.collect(Collectors.toList());
-	}
-
-	private void registerOpsFrom(final Map<String, List<OpInfo>> opDirectory, final Object o, List<OpInfoGenerator> generators) {
-		opsFromObject(o, generators).forEach(info -> addToOpIndex.accept(opDirectory, info));
-	}
-	
-	private List<OpInfoGenerator> infoGenerators() {
-		return discoverers.stream() //
-				.flatMap(d -> d.discover(OpInfoGenerator.class).stream()) //
-				.collect(Collectors.toList());
-	}
-
 	private synchronized void initIdDirectory() {
 		if (idDirectory != null) return;
 		idDirectory = new HashMap<>();
-		if (opDirectory == null) initOpDirectory();
 
 		opDirectory.values().stream() //
 				.flatMap(Collection::stream) //
 				.forEach(info -> idDirectory.put(info.id(), info));
 	}
 
-	private final BiConsumer<Map<String, List<OpInfo>>, OpInfo> addToOpIndex = (final Map<String, List<OpInfo>> directory, final OpInfo opInfo) -> {
+	private final Consumer<OpInfo> addToOpIndex = (final OpInfo opInfo) -> {
 		if (opInfo.names() == null || opInfo.names().isEmpty()) {
 			log.error("Skipping Op " + opInfo.implementationName() + ":\n" +
 				"Op implementation must provide name.");
 			return;
 		}
 		for (String opName : opInfo.names()) {
-			directory.computeIfAbsent(opName, name -> new ArrayList<>()).add(opInfo);
+			opDirectory.computeIfAbsent(opName, name -> new TreeSet<>()).add(opInfo);
```

Comment:
> Is the use of `TreeSet` so that it always stays sorted?

@ctrueden yes, I want it sorted by Op priority - I'm starting to think that maybe instead of `List<OpInfo>` being returned by `OpEnvironment.infos` we should return a `SortedSet<OpInfo>` - that way, you'd get two guarantees:
1. The sorting gives you Ops by priority
2. The set prevents duplicates.

Reply:
Yes, `SortedSet` seems like the right interface here! :+1:

Reply:
Changed things to use `SortedSet` across the board.